lib/win32/certstore/store_base.rb

```diff
@@ -0,0 +1,95 @@
+# limitations under the License.
+
+require 'mixin/crypto'
+require 'openssl'
+
+module Win32
+  class Certstore
+    module StoreBase
+      include Win32::Mixin::Crypto
+      include Win32::Mixin::Assertions
+      include Chef::Mixin::WideString
+      include Chef::Mixin::ShellOut
+
+      def cert_list(store_handler)
+        cert_name = FFI::MemoryPointer.new(2, 128)
+        cert_list = []
+        begin
+          while (pCertContext = CertEnumCertificatesInStore(store_handler, pCertContext) and not pCertContext.null? ) do
+            if (CertGetNameStringW(pCertContext, CERT_NAME_FRIENDLY_DISPLAY_TYPE, CERT_NAME_ISSUER_FLAG, nil, cert_name, 1024))
+              cert_list << cert_name.read_wstring
+            end
+          end
+          CertFreeCertificateContext(pCertContext)
+        rescue Exception => e
+          lookup_error("list")
+        end
+        cert_list.to_json
+      end
+      
+      def cert_add(store_handler, cert_file_path)
+        validate_certificate(cert_file_path)
+        file_content = read_certificate_content(cert_file_path)
+        pointer_cert = FFI::MemoryPointer.from_string(file_content)
+        cert_length = file_content.bytesize
+        begin
+          if (CertAddEncodedCertificateToStore(store_handler, X509_ASN_ENCODING, pointer_cert, cert_length, 2, nil))
+            "Added certificate #{File.basename(cert_file_path)} successfully"
+          else
+            lookup_error
+          end
+        rescue Exception => e
+          lookup_error("add")
+        end
+      end
+
+      private
+
+      def lookup_error(failed_operation = nil)
+        last_error = FFI::LastError.error
+        case last_error
+        when 1223
+          raise Chef::Exceptions::Win32APIError, "The operation was canceled by the user."
+        when -2146885628
+          raise Chef::Exceptions::Win32APIError, "Cannot find object or property."
+        when -2146885629
+          raise Chef::Exceptions::Win32APIError, "An error occurred while reading or writing to a file."
+        when -2146881269
+          raise Chef::Exceptions::Win32APIError, "ASN1 bad tag value met. -- Is the certificate in DER format?"
+        when -2146881278
+          raise Chef::Exceptions::Win32APIError, "ASN1 unexpected end of data."
+        else
+          raise Chef::Exceptions::Win32APIError, "Unable to #{failed_operation} certificate with error: #{last_error}."
+        end
+      end
+
+      # This is a single public certificate in X509 DER format.
+      # If your certificate has a header and footer line like "---- BEGIN CERTIFICATE ----" then it is in PEM format, not DER format.
+      # A certificate can be converted with `openssl x509 -in example.crt -out example.der -outform DER`
+      def read_certificate_content(cert_path)
+        unless (File.extname(cert_path) == ".der")
+          temp_file = shell_out("powershell.exe -Command $env:temp").stdout.strip.concat("\\TempCert.der")
```

Comment:
Use `Tempfile.new` here. This will:

1) use the temp directory without having to shell out to get it (using `ENV['TEMP']` in Ruby would have been more appropriate here)
2) produce a temp file with a random string inserted so you don't have to worry about the existence of a file, or a security issue with someone maliciously creating a temporary file with that filename, etc.